Migrate zk asset tests

thomas-waite opened the pull request:
## Description
Migrate `ZkAsset.js`, `ZkAssetMintable.js`, `ZkAssetBurnable.js` and `ZkAssetOwnable.js` tests to the new proof construction. 

<!--- Describe your changes in detail -->

## Testing instructions
`yarn test`
<!--- Please describe how reviewers can test your changes -->

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

<!-- * New feature (non-breaking change which adds functionality) -->

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.

PaulRBerg committed:
refactor: prune leftovers from rebase and remove sigs from join-split

thomas-waite committed:
test(protocol): migrate to new proof construction in ZkAsset tests
migrate tests for ZkAsset, ZkAssetMintable, ZkAssetBurnable

thomas-waite committed:
feat(aztec.js): add method to construct ECDSA signatures for joinSplit

thomas-waite committed:
test(protocol): finish migrating ZkAsset tests
implement tests for ZkAssetOwnable
eslint and prettier fixes

thomas-waite committed:
refactor: remove constructOutput() joinSplit proof method
eslint fixes in protocol package
remove .only

PaulRBerg commented on packages/aztec.js/src/proof/joinSplit/index.js:
```diff
@@ -127,6 +128,32 @@ class JoinSplitProof extends Proof {
+     * @param {string[]} aztecAccounts mapping between owners and private keys
+     * @returns {string} array of signatures
+     */
+    constructSignatures(validatorAddress, inputNoteOwners) {
```
I suggest rewriting this as a mapping instead of an array. Find an example here: https://github.com/AztecProtocol/AZTEC/blob/0d5af182ca4d7698654ffbfd348f02a555464d94/packages/aztec.js/src/proof/joinSplit/index.js#L131

You can merge over `this.inputNotes` instead of `inputNoteOwners`.

Aslo, make sure the comments are in sync with the parameter names.

thomas-waite replied:
As discussed, leaving for now

PaulRBerg commented on packages/protocol/test/ERC1724/ZkAsset.js:
```diff
@@ -1,52 +1,93 @@
+
+const setupSingleProofTest = async (noteValues) => {
+    const numNotes = noteValues.length;
+    const aztecAccounts = [...new Array(numNotes)].map(() => secp256k1.generateAccount());
```
One thing that I did and was highly effective in my refactored tests was to get rid of `aztecAccounts` and notes. That is, simply have one or two accounts and a few notes that have bespoke names.

thomas-waite replied:
Good ideas, refactored these setup functions and updated tests

PaulRBerg commented on packages/protocol/test/ERC1724/ZkAsset.js:
```diff
@@ -1,52 +1,93 @@
+    };
+};
+
+const setupTwoProofTest = async (noteValues) => {
```
These functions should have names that are more specific.

thomas-waite replied:
Good point, done

PaulRBerg commented on packages/protocol/test/ERC1724/ZkAsset.js:
```diff
@@ -126,394 +157,383 @@ contract.skip('ZkAsset', (accounts) => {
+            await ace.publicApprove(zkAsset.address, depositProof.hash, depositPublicValue, { from: accounts[0] });
+            await zkAsset.confidentialTransfer(depositData, depositSignatures);
+
+            const fakeInputNoteOwners = [secp256k1.generateAccount(), secp256k1.generateAccount()];
```
In my refactor, the standard naming for this kind of test variables switched to `malformed` instead of `fake`.

thomas-waite replied:
Updated the names

PaulRBerg commented on packages/protocol/test/ERC1724/ZkAssetMintable.js:
```diff
@@ -1,264 +1,228 @@
+                canAdjustSupply,
+                canConvert,
+                {
+                    from: accounts[0],
```
Not necessary to specify `from` here.

thomas-waite replied:
Agreed, but in some cases the address of the sender is significant. A proof may fail if an incorrect sender address was used, so I've taken to making the `from` field explicit

PaulRBerg commented on packages/protocol/test/ERC1724/ZkAssetOwnable.js:
```diff
@@ -358,155 +400,147 @@ contract.skip('ZkAssetOwnable', (accounts) => {
-            const formattedProofOutput = `0x${transferProofOutput.slice(0x40)}`;
             await truffleAssert.reverts(
-                zkAssetOwnableTest.callConfidentialTransferFrom(JOIN_SPLIT_PROOF, formattedProofOutput),
+                zkAssetOwnableTest.callConfidentialTransferFrom(JOIN_SPLIT_PROOF, transferProof.eth.output),
```
Double-check that this is not a false positive. Now there's both `proof.eth.output` and `proof.eth.outputs`.

thomas-waite replied:
Done

thomas-waite committed:
test(protocol): implement PR feedback + more scenario style testing
created a helper/ERC1724.js file, which sets up scenario like testing situations for ZkAsset tests
rename some variables to be consistent with new standardised format

thomas-waite committed:
test(protocol): remove unnecessary publicApprove() calls

remove unnecessary publicApprove() calls in tests for transfer of tokens out of ACE. remove
redundant failure tests for publicApprove() on transferProof in ZkAssetOwnable.js

thomas-waite committed:
test(protocol): fix validator tests variable integration into challenge

tests were returning a false positive. Needed to reset the proof.challengeHash to an empty
instantiation of the Keccak class, before calling the constructChallengeRecurse() method. This is
needed because when the proof class is instantiated, constructChallengeRecurse() is automatically
called - so proof.challengeHash will already have been hashed
